### PR TITLE
feat: Complete rpc client implementation using a generator

### PR DESCRIPTION
**Pull request by iharobCP**

### CODE REVIEW

**Comment by ChikinDeveloper** [comment]
Hi! If I'm right, with this implementation of Filter you have to filter by memcmp and dataSize, you can't filter by dataSize only (or memcmp only) in getProgramAccounts.

This is how I would implement it

```dart
abstract class Filter {
  Map<String, dynamic> toJson();
}

@JsonSerializable(createFactory: true, includeIfNull: false)
class DataSizeFilter implements Filter {
  const DataSizeFilter({required this.dataSize});

  factory DataSizeFilter.fromJson(Map<String, dynamic> json) =>
      _$DataSizeFilterFromJson(json);

  @override
  Map<String, dynamic> toJson() => _$DataSizeFilterToJson(this);

  /// Compares the program account data length with the provided
  /// data size.
  final int dataSize;
}

@JsonSerializable(createFactory: true, includeIfNull: false)
class MemcmpFilter implements Filter {
  const MemcmpFilter({
    required this.offset,
    required this.bytes,
  });

  factory MemcmpFilter.fromJson(Map<String, dynamic> json) =>
      _$MemcmpFilterFromJson(json);

  @override
  Map<String, dynamic> toJson() => _$MemcmpFilterToJson(this);

  /// Offset into program account data to start comparison.
  final int offset;

  /// Data to match, as base-58 encoded string and limited to
  /// less than 129 bytes.
  final String bytes;
}
```

Not really needed but you can also check that there is always at most 1 DataSizeFilter in a list, and check Memcmp filters conflicts.

File: packages/solana/lib/src/rpc/dto/account_data/account_data.dart [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      return ParsedAccountData.fromJson(data);
+    } else if (data is String) {
+      return AccountData._fromBase64String(data);
+    } else {
+      throw const FormatException('account data is in unknown format');
+    }
+  }
+
+  Map<String, dynamic> toJson() {
+    throw UnsupportedError('converting account data to json is not supported');
```
Comment (ookami-kb) [inline packages/solana/lib/src/rpc/dto/account_data/account_data.dart:73]: If this is not supported, let's remove this method at all.

File: packages/solana/lib/src/rpc/dto/account_data/spl_token_program/mint_account_data_info.dart [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import 'package:json_annotation/json_annotation.dart';
+
+part 'mint_account_data_info.g.dart';
+
+@JsonSerializable()
```
Comment (ookami-kb) [inline packages/solana/lib/src/rpc/dto/account_data/spl_token_program/mint_account_data_info.dart:5]: `toJson` is not used, should we remove it from generation? Same for other classes, I think for responses it doesn't make sense to keep it.

File: packages/solana/lib/src/rpc/dto/account_data/spl_token_program/token_program_account_data.dart [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    required String type,
+  }) = UnknownAccountData;
+
+  factory SplTokenProgramAccountData.fromJson(
+    Map<String, dynamic> data,
+  ) =>
+      _$SplTokenProgramAccountDataFromJson(data);
+
+  Map<String, dynamic> toJson() {
+    throw const FormatException('not serializable');
```
Comment (ookami-kb) [inline packages/solana/lib/src/rpc/dto/account_data/spl_token_program/token_program_account_data.dart:32]: Why? It's better to remove it at all.

File: packages/solana/lib/src/rpc/dto/account_data/empty_account_data.dart [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import 'package:solana/src/rpc/dto/account_data/account_data.dart';
+
+class EmptyAccountData implements AccountData {
```
Comment (ookami-kb) [inline packages/solana/lib/src/rpc/dto/account_data/empty_account_data.dart:3]: This is not used anywhere.

File: packages/solana/lib/src/rpc/dto/account_data/parsed_account_data.dart [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  const factory ParsedAccountData.unsupported(
+    Map<String, dynamic> parsed,
+  ) = UnsupportedProgramAccountData;
+
+  factory ParsedAccountData.fromJson(Map<String, dynamic> json) =>
+      _$ParsedAccountDataFromJson(json);
+
+  const ParsedAccountData._();
+
+  dynamic get info => map<dynamic>(
```
Comment (ookami-kb) [inline packages/solana/lib/src/rpc/dto/account_data/parsed_account_data.dart:29]: I would remove this method at all, it's pretty confusing. If needed, one can do the manul mapping.

File: packages/solana/lib/src/rpc/dto/account_data/saved.g.dart [hunk 6]
```diff
@@ -0,0 +1,24 @@
+// GENERATED CODE - DO NOT MODIFY BY HAND
```
Comment (ookami-kb) [inline packages/solana/lib/src/rpc/dto/account_data/saved.g.dart:1]: Is this file needed at all?

File: packages/solana/lib/src/rpc/dto/parsed_message/header.dart [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+  factory Header.fromJson(dynamic json) {
+    if (json is List) {
+      return Header(
+        numRequiredSignatures: json[0] as int,
+        numReadonlyUnsignedAccounts: json[1] as int,
+        numReadonlySignedAccounts: json[2] as int,
+      );
+    } else {
+      return Header(
```
Comment (ookami-kb) [inline packages/solana/lib/src/rpc/dto/parsed_message/header.dart:17]: You can combine manual and generated code:

```dart
@JsonSerializable(createToJson: false)
class Header {
  Header({
    required this.numRequiredSignatures,
    required this.numReadonlySignedAccounts,
    required this.numReadonlyUnsignedAccounts,
  });

  factory Header.fromJson(dynamic json) {
    if (json is List) {
      return Header(
        numRequiredSignatures: json[0] as int,
        numReadonlyUnsignedAccounts: json[1] as int,
        numReadonlySignedAccounts: json[2] as int,
      );
    } else {
      return _$HeaderFromJson(json as Map<String, dynamic>);
    }
  }

  final int numRequiredSignatures;
  final int numReadonlySignedAccounts;
  final int numReadonlyUnsignedAccounts;
}
```

File: packages/solana/lib/src/rpc/dto/block_commitment.dart [hunk 8]
```diff
@@ -0,0 +1,25 @@
+import 'package:json_annotation/json_annotation.dart';
+
+part 'block_commitment.g.dart';
+
+/// The commitment of a bloc
+@JsonSerializable(createFactory: true, includeIfNull: false)
```
Comment (ookami-kb) [inline packages/solana/lib/src/rpc/dto/block_commitment.dart:6]: `createFactory` is `true` by default.

**Review by ookami-kb — CHANGES_REQUESTED** [review]
Review is still in progress, publishing intermediate comments.

File: packages/solana/lib/src/rpc/dto/account_data/account_data.dart (hunk 1 above)
Comment (iharobCP) [reply packages/solana/lib/src/rpc/dto/account_data/account_data.dart:73]: Good catch. It is here only because when using freezed it was required. But no longer because we are using `json_serializable`

File: packages/solana/lib/src/rpc/dto/account_data/spl_token_program/token_program_account_data.dart (hunk 3 above)
Comment (iharobCP) [reply packages/solana/lib/src/rpc/dto/account_data/spl_token_program/token_program_account_data.dart:32]: Same reason. It was here for `@freezed`

File: packages/solana/lib/src/rpc/dto/account_data/empty_account_data.dart (hunk 4 above)
Comment (iharobCP) [reply packages/solana/lib/src/rpc/dto/account_data/empty_account_data.dart:3]: That's a mistake. It should be.

File: packages/solana/lib/src/rpc/dto/account_data/parsed_account_data.dart (hunk 5 above)
Comment (iharobCP) [reply packages/solana/lib/src/rpc/dto/account_data/parsed_account_data.dart:29]: I agree.

File: packages/solana/lib/src/rpc/dto/account_data/saved.g.dart (hunk 6 above)
Comment (iharobCP) [reply packages/solana/lib/src/rpc/dto/account_data/saved.g.dart:1]: Not at all. It was a temporary file. How did you see it? :fire:

File: packages/solana/lib/src/rpc/dto/parsed_message/header.dart (hunk 7 above)
Comment (iharobCP) [reply packages/solana/lib/src/rpc/dto/parsed_message/header.dart:17]: I love it ;)

File: packages/solana/lib/src/rpc/dto/account_data/empty_account_data.dart (hunk 4 above)
Comment (iharobCP) [reply packages/solana/lib/src/rpc/dto/account_data/empty_account_data.dart:3]: It is used, in fact. I checked.

File: packages/jsonrpc_client/lib/src/client_generator.dart [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+      final fromJsonParameters = [
+        '$data as Map<String, dynamic>',
+        if (genericFactory.isNotEmpty) genericFactory,
+      ];
+
+      if (typeIsPrimitive) {
+        return '$nullCheck$data as $typeName';
+      }
+
+      return '$nullCheck$typeName.fromJson(${fromJsonParameters.join(', ')})';
```
Comment (ookami-kb) [inline packages/jsonrpc_client/lib/src/client_generator.dart:149]:
```suggestion
      final type = getDisplayString(withNullability: false);
      return '$nullCheck$type.fromJson(${fromJsonParameters.join(', ')})';
```

**Review by ookami-kb — CHANGES_REQUESTED** [review]

File: packages/jsonrpc_client/lib/src/client_generator.dart [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+  String _generateMethod(MethodElement method) {
+    final params = method.parameters.where((p) => p.isPositional).map((p) {
+      if (p.type.isNullableType) {
+        return 'if (${p.name} != null) ${p.toJson()}';
+      } else {
+        return p.toJson();
+      }
+    }).toList();
+    final isWithContext = method.metadata.any(
+      (a) => a.toSource() == "@withContext",
```
[the remark this answers is not in the document]
Comment (iharobCP) [reply packages/jsonrpc_client/lib/src/client_generator.dart:60]: I didn't know this. Thanks.

File: analysis_options.yaml [hunk 11]
```diff
@@ -1 +1,20 @@
 include: package:mews_pedantic/analysis_options.yaml
+
+dart_code_metrics:
```
Comment (ookami-kb) [inline analysis_options.yaml:3]: This can be removed as well.

**Review by ookami-kb — COMMENTED** [review]
One comment, otherwise LGTM.

**Review by ookami-kb — APPROVED** [review]
